Flavor, type and spec changes

w-miller opened the pull request:
* Extend specs to contain a BMC type
* Make Ironic enrolment optional with node type ironic_config attribute
* Add Nova flavor creation

Commit:
Remove properties from Ironic nodes

These are no longer used, in favour of resource classes.

Commit:
Make specs, types and flavors more extensible

* Give specs an optional Ironic config field
* Add bmc_type field to attempt to decouple IPMI
* Add nova_flavors to allow flavor creation

Commit:
Save ironic_config in node details

Commit:
Add bmc_type checking for virtualbmc

Commit:
Use node ironic_config details during enrolment

Commit:
Add Nova flavor registration

Commit:
Add Ansible structure diagram to README

markgoddard commented on README.md:
```diff
@@ -15,4 +15,18 @@ installed inside it, Tenks' role dependencies can be installed by
+rectangles represent playbooks, green rounded rectangles represent task books,
+red ellipses represent roles and yellow rhombi represent action plugins.
+
+![Tenks Ansible structure](assets/tenks_ansible_structure.png)
```
As helpful as this is, it could become stale. How would one go about updating it?

w-miller replied:
That is true. Ideally I'd write a tool that auto-generates this sort of diagram from a pile of Ansible, so we could dynamically regenerate the image artifact in Tox... or something.

As it stands, I'd have to export it from draw.io and commit the new PNG. Can you think of a better way to keep this diagram fresh here? Or perhaps it only has a place in the blog post.

markgoddard replied:
If you share the diagram on draw.io, maybe add a comment with a link to it, then it can be updated when you're gone.

markgoddard commented on ansible/action_plugins/tenks_schedule.py:
```diff
@@ -48,16 +47,32 @@ def run(self, tmp=None, task_vars=None):
             for _ in six.moves.range(cnt):
                 node = deepcopy(task_vars['node_types'][typ])
+                # All nodes need a BMC type.
+                node.setdefault('bmc_type', task_vars['hostvars']['localhost'][
```
would it make sense to use driver/hardware_type here to be consistent with ironic? I can't decide.

w-miller replied:
At the moment, we've got a default hardware_type specified for each bmc_type in default_bmc_info. If you think the bmc_types we have will always map to existing Ironic hardware types (drivers), then we could remove this extra layer.

markgoddard replied:
Let's leave it as is for now.

w-miller replied:
I've got a patch ready to use driver names instead - probably worth using now I've done it, right?

markgoddard replied:
Sure

markgoddard commented on ansible/host_vars/localhost:
```diff
@@ -21,19 +18,89 @@ allocations_file_path: >-
+#       hardware_type: ipmi
+#       boot_interface: pxe
+#       deploy_interface: iscsi
+#       management_interface: ipmitool
```
Also: network_interface, bios_interface, raid_interface, console_interface, storage_interface, inspect_interface, vendor_interface

w-miller replied:
Should I set defaults for all of those here for IPMI and Redfish?

markgoddard replied:
I'm wondering whether duplicating the mappings in ironic is the right idea. How about we keep it configurable per spec, but allow them to be omitted and use ironic's defaults?

w-miller replied:
Sure, sounds like a better idea to me.

markgoddard commented on ansible/host_vars/localhost:
```diff
@@ -21,19 +18,89 @@ allocations_file_path: >-
+#       traits: []
+#       # The following variables are all optional; if not set, a default will
+#       # be inferred from the `bmc_type` specified by the node type `type`.
+#       hardware_type: ipmi
```
Let's call this 'driver', that how ironic's API refers to it

w-miller replied:
Sure.

markgoddard commented on ansible/host_vars/localhost:
```diff
@@ -21,19 +18,89 @@ allocations_file_path: >-
+#       # Required.
+#       resource_class: my_rc
+#       # Optional - defaults to [].
+#       traits: []
```
Also: Capabilities, or properties.

w-miller replied:
Ok.

markgoddard commented on ansible/host_vars/localhost:
```diff
@@ -21,19 +18,89 @@ allocations_file_path: >-
+# The default BMC type of a node. Can be overridden per-node.
+default_bmc_type: ipmi
+# Maps BMC types to a dict containing default values for Ironic node
+# interfaces, a hardware type, and a BMC emulation tool to use. Any of these
```
delete: and a BMC emulation tool to use?

w-miller replied:
Yep, that's irrelevant now. Thanks.

markgoddard commented on ansible/register_flavors.yml:
```diff
@@ -0,0 +1,11 @@
+  tasks:
+    - name: Register Nova flavors
+      include_role:
+        name: nova-flavors
```
Sorry, should have mentioned stackhpc.os-flavors. This is fine.

w-miller replied:
I did find stackhpc.os-flavors but it seemed to only really added a wrapper around the ansible_python_interpreter setting/unsetting, which we don't need here anyway, and I'd still need to do the extra_specs fiddling, so I didn't think it would add much benefit.

w-miller commented:
Closing and re-opening to trigger Travis.

markgoddard commented on ansible/roles/ironic-enrolment/tasks/node.yml:
```diff
@@ -61,18 +63,30 @@
+               default_bmc_info[node.bmc_type].management_interface) }}
+        --power-interface
+          {{ node.ironic_config.power_interface | default(
+               default_bmc_info[node.bmc_type].power_interface) }}
```
Could potentially do this in a loop

markgoddard commented on ansible/roles/nova-flavors/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+             }
+             | combine(dict(item.required_traits | default([])
+                       | map('regex_replace', '(.*)', 'trait:\1')
+                       | zip_longest([], fillvalue='required')))
```
Inventive!

markgoddard commented on ansible/roles/nova-flavors/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+        #   * Add the resource class name.
+        #   * Add required and forbidden traits.
+        #   * Add any custom specs from the user.
+        extra_specs: >-
```
This might be easier on the eye if each of the dicts being combined were defined as a task variable

w-miller replied:
Ok.

w-miller replied:
Sure.

markgoddard commented on ansible/roles/nova-flavors/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+                  | size_string_to_gb }}"
+        ram: "{{ node_types[item.node_type].memory_mb }}"
+        vcpus: "{{ node_types[item.node_type].vcpus }}"
+        # NOTE(w-miller): I'm not quite sure whether this is janky or beautiful.
```
Can it not be both?

w-miller replied:
I hope so!

Commit:
Add diagram update instructions

Commit:
Remove 'bmc type' concept

We can directly refer to a node's Ironic driver rather than having an
abstraction for a BMC type.

Commit:
Add defaults for all Ironic interfaces

markgoddard commented:
Merge conflicts

Commit:
Set interfaces and properties on Ironic nodes

Pass any interface that was specified through to Ironic; otherwise, let
Ironic decide a default for it. Also, pass through any property
key-value pairs we get.

Commit:
Extract Ironic node extra specs dicts

It improves readability to have each dict defined separately.

Commit:
Merge branch 'master' into resource-classes

Conflicts:
	ansible/action_plugins/tenks_schedule.py

w-miller commented:
Travis' apt package installation doesn't seem to be deterministic - it fell over with `no installation candidates for 'python-virtualenv'` and `no package named 'realpath'`, but restarting the build without changes made it pass.

Updated and ready for re-review.
